examples/sampling.py

```diff
@@ -0,0 +1,247 @@
+
+
+def plot(frames: torch.Tensor, title : Optional[str] = None):
+    try:
+        from torchvision.utils import make_grid
+        from torchvision.transforms.v2.functional import to_pil_image
+        import matplotlib.pyplot as plt
+    except ImportError:
+        print("Cannot plot, please run `pip install torchvision matplotlib`")
+        return
+
+    plt.rcParams["savefig.bbox"] = 'tight'
+    fig, ax = plt.subplots()
+    ax.imshow(to_pil_image(make_grid(frames)))
+    ax.set(xticklabels=[], yticklabels=[], xticks=[], yticks=[])
+    if title is not None:
+        ax.set_title(title)
+    plt.tight_layout()
+
+
+# %%
+# .. _sampling_tuto_start:
+#
+# Creating a decoder
+# ------------------
+#
+# Sampling clips from a video always starts by creating a
+# :class:`~torchcodec.decoders.VideoDecoder` object. If you're not already
+# familiar with :class:`~torchcodec.decoders.VideoDecoder`, take a quick look
+# at: :ref:`sphx_glr_generated_examples_basic_example.py`.
+from torchcodec.decoders import VideoDecoder
+
+# You can also pass a path to a local file!
+decoder = VideoDecoder(raw_video_bytes)
+
+# %%
+# Sampling basics
+# ---------------
+#
+# We can now use our decoder to sample clips. Let's first look at a simple
+# example: all other samplers follow similar APIs and principles. We'll use
+# :func:`~torchcodec.samplers.clips_at_random_indices` to sample clips that
+# start at random indices.
+
+from torchcodec.samplers import clips_at_random_indices
+
+torch.manual_seed(0)  # The samplers RNG is controlled by pytorch's RNG
+clips = clips_at_random_indices(
+    decoder,
+    num_clips=5,
+    num_frames_per_clip=4,
+    num_indices_between_frames=3,
+)
+clips
+
+# %%
+# The output of the sampler is a sequence of clips, represented as
+# :class:`~torchcodec.FrameBatch` object. In this object, we have different
+# fields:
+#
+# - ``data``: a 5D uint8 tensor representing the frame data. Its shape is
+#   (num_clips, num_frames_per_clip, ...) where ... is either (C, H, W) or (H,
+#   W, C), depending on the ``dimension_order`` parameter of the
+#   :class:`~torchcodec.VideoDecoder`. This is typically what would get passed
+#   to the model.
+# - ``pts_seconds``: a 2D float tensor of shape (num_clips, num_frames_per_clip)
+#   giving the starting timestamps of each frame within each clip, in seconds.
+# - ``duration_seconds``: a 2D float tensor of shape (num_clips,
+#   num_frames_per_clip) giving the duration of each frame within each clip, in
+#   seconds.
+
+plot(clips[0].data)
+
+# %%
+# Indexing and manipulating clips
+# -------------------------------
+#
+# Clips are :class:`~torchcodec.FrameBatch` objects, and they support native
+# pytorch indexing semantics (including fancy indexing). This makes it easy to
+# filter clips based on a given criteria. For example, from the clips above we
+# can easily filter out those who start *after* a specific timestamp:
+clip_starts = clips.pts_seconds[:, 0]
```

Comment:
I actually don't fully understand what we're doing here - I clearly have a lot to learn about PyTorch tensors. :) Are we making a view (copy?) of all `clip.pts_seconds`? Can you explain (to me, not necessarily to the reader) the mechanics of what's going on?

Reply:
I think pytorch is able to create a view, because with that index the strides are regular. I don't know if there's any such guarantee in general though.

`[:, i]`  is a typical indexing strat that means "select the i'th elements along the 2nd dim, don't index the first dim". We end up with a tensor with the same length as the first dim. There can be as many `:` as there are dimension. PyTorch largely follows numpy's indexing semantics, and they do a much better job that I will at explaning all this: https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing